"invalid initializer" C error when iterating on a generic map of string T from a generic function

### Describe the bug

When my function declares a first parameter as a map of string generic, there is a C error. When I remove the generic, no error is generated and the program terminates as expected.

### Reproduction Steps

Using the following code snippet there is a C error:

```v
module main

fn validate[T](values map[string]T, rules map[string][]string) ! {
  for key, rule_list in rules {
    value := values[key]! // <------------------
  }
}

fn main() {
  validate({
    "age": 31
  }, {
    "age": [
      "required"
    ]
  }) or {
    eprintln(err.msg())
  }
}
```

### Expected Behavior

The program terminates without errors.

### Current Behavior

It displays this C error stack trace:

```
================== C compilation error (from gcc): ==============
cc: /tmp/v_0/main.01JGSHJTQA2ZR9TN1W7PHK9DRV.tmp.c: In function 'main__validate_T_int':
cc: /tmp/v_0/main.01JGSHJTQA2ZR9TN1W7PHK9DRV.tmp.c:7086:26: error: invalid initializer
cc:  7086 |   Map_string_int value = (*(int*)_t4.data);
cc:       |                          ^
=================================================================
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.4.9

### Environment details (OS name and version, etc.)

V full version: V 0.4.9 4225a34
OS: linux, N/A
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz

getwd: /home/v
vexe: /opt/v/v
vexe mtime: 2024-12-30 03:28:29

vroot: OK, value: /opt/v
VMODULES: OK, value: /root/.vmodules
VTMP: OK, value: /tmp/v_0

env VFLAGS: "-cc gcc"

Git version: git version 2.30.6
Git vroot status: 4225a346 (44 commit(s) behind V master)
.git/config present: true

CC version: cc (Alpine 10.2.1_pre1) 10.2.1 20201203
emcc version: N/A
thirdparty/tcc status: thirdparty-linuxmusl-amd64 a3e24da2

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2Nzc5YTA5NmYxNDRmNTg1YWU1MjgxZjAiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.qyKqm60AijteaU08KdDoE6UYz1ipL0TdcIzpmgG5RCQ">Huly&reg;: <b>V_0.6-21807</b></a></sub>
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -341,15 +341,13 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 						}
 						g.assign_ct_type = var_type
 					}
-				} else if val is ast.IndexExpr {
-					if val.left is ast.Ident && g.type_resolver.is_generic_param_var(val.left) {
-						ctyp := g.unwrap_generic(g.get_gn_var_type(val.left))
-						if ctyp != ast.void_type {
-							var_type = ctyp
-							val_type = var_type
-							left.obj.typ = var_type
-							g.assign_ct_type = var_type
-						}
+				} else if val is ast.IndexExpr && (val.left is ast.Ident && val.left.ct_expr) {
+					ctyp := g.unwrap_generic(g.type_resolver.get_type(val))
+					if ctyp != ast.void_type {
+						var_type = ctyp
+						val_type = var_type
+						left.obj.typ = var_type
+						g.assign_ct_type = var_type
 					}
 				} else if left.obj.ct_type_var == .generic_var && val is ast.CallExpr {
 					if val.return_type_generic != 0 && val.return_type_generic.has_flag(.generic) {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1353,27 +1353,6 @@ fn (mut g Gen) gen_to_str_method_call(node ast.CallExpr) bool {
 	return false
 }
 
-fn (mut g Gen) get_gn_var_type(var ast.Ident) ast.Type {
-	if g.cur_fn != unsafe { nil } && g.cur_fn.generic_names.len > 0 {
-		for k, cur_param in g.cur_fn.params {
-			if (k == 0 && g.cur_fn.is_method) || !cur_param.typ.has_flag(.generic)
-				|| var.name != cur_param.name {
-				continue
-			}
-			mut typ := cur_param.typ
-			mut cparam_type_sym := g.table.sym(g.unwrap_generic(typ))
-
-			if cparam_type_sym.kind == .array {
-				typ = g.unwrap_generic((cparam_type_sym.info as ast.Array).elem_type)
-			} else if cparam_type_sym.kind == .array_fixed {
-				typ = g.unwrap_generic((cparam_type_sym.info as ast.ArrayFixed).elem_type)
-			}
-			return typ
-		}
-	}
-	return ast.void_type
-}
-
 // resolve_return_type resolves the generic return type of CallExpr
 fn (mut g Gen) resolve_return_type(node ast.CallExpr) ast.Type {
 	if node.is_method {

--- a/vlib/v/tests/comptime/comptime_generic_map_test.v
+++ b/vlib/v/tests/comptime/comptime_generic_map_test.v
@@ -1,0 +1,27 @@
+module main
+
+fn validate[T](values map[string]T, rules map[string][]string) ! {
+	for key, _ in rules {
+		value := values[key]!
+		assert typeof(value).name == T.name
+	}
+}
+
+fn test_main() {
+	validate({
+		'age': 31
+	}, {
+		'age': [
+			'required',
+		]
+	}) or { assert false }
+
+	validate({
+		'foo': 'bar'
+	}, {
+		'foo': [
+			'required',
+		]
+	}) or { assert false }
+	assert true
+}

